core/src/main/java/org/apache/iceberg/FindFiles.java

```diff
@@ -0,0 +1,195 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+import com.google.common.base.Preconditions;
+import java.time.Instant;
+import java.time.LocalDateTime;
+import java.time.ZoneId;
+import java.time.format.DateTimeFormatter;
+import java.util.Arrays;
+import java.util.List;
+import org.apache.iceberg.expressions.Expression;
+import org.apache.iceberg.expressions.Expressions;
+import org.apache.iceberg.io.CloseableIterable;
+
+public class FindFiles {
+  private static final DateTimeFormatter DATE_FORMAT = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss.SSS");
+
+  public static Builder in(Table table) {
+    return new Builder(table);
+  }
+
+  public static class Builder {
+    private final Table table;
+    private final TableOperations ops;
+    private boolean caseSensitive = true;
+    private Long snapshotId = null;
+    private Expression rowFilter = Expressions.alwaysTrue();
+    private Expression fileFilter = Expressions.alwaysTrue();
+    private Expression partitionFilter = Expressions.alwaysTrue();
+
+    public Builder(Table table) {
+      this.table = table;
+      this.ops = ((HasTableOperations) table).operations();
+    }
+
+    public Builder caseInsensitive() {
+      this.caseSensitive = false;
+      return this;
+    }
+
+    public Builder caseSensitive(boolean caseSensitive) {
+      this.caseSensitive = caseSensitive;
+      return this;
+    }
+
+    /**
+     * Base results on the given snapshot.
+     *
+     * @param snapshotId a snapshot ID
+     * @return this for method chaining
+     */
+    public Builder inSnapshot(long snapshotId) {
+      Preconditions.checkArgument(this.snapshotId == null,
+          "Cannot set snapshot multiple times, already set to id=%s", snapshotId);
+      Preconditions.checkArgument(table.snapshot(snapshotId) != null,
+          "Cannot find snapshot for id=%s", snapshotId);
+      this.snapshotId = snapshotId;
+      return this;
+    }
+
+    /**
+     * Base results on files in the snapshot that was current as of a timestamp.
+     *
+     * @param timestampMillis a timestamp in milliseconds
+     * @return this for method chaining
+     */
+    public Builder asOfTime(long timestampMillis) {
+      Preconditions.checkArgument(this.snapshotId == null,
+          "Cannot set snapshot multiple times, already set to id=%s", snapshotId);
+
+      Long lastSnapshotId = null;
+      for (HistoryEntry logEntry : ops.current().snapshotLog()) {
+        if (logEntry.timestampMillis() <= timestampMillis) {
+          lastSnapshotId = logEntry.snapshotId();
+        }
```

Comment:
I assume `snapshotLog()` is in ascending order. Should we then `break` after the condition is not true anymore?

Reply:
Done, and I added a test for `asOfTime`.

The log is in ascending order and that is validated here: https://github.com/apache/incubator-iceberg/blob/master/core/src/main/java/org/apache/iceberg/TableMetadata.java#L178-L185

Reply:
If its ascending order, and we want the closest to timestamp, then shouldn't it be:
```
if (logEntry.timestampMillis() <= timestampMillis) {
  lastSnapshotId = logEntry.snapshotId();
} else {
  break;
}
```

In other words, I think this test would fail right now:

```
          @Test
	  public void testAsOfTimestamp() {
	    table.newAppend()
	        .appendFile(FILE_A)
	        .commit();

	    table.newAppend()
	        .appendFile(FILE_B)
	        .commit();
	
	    long timestamp = System.currentTimeMillis();
	
	    table.newAppend()
	        .appendFile(FILE_C)
	        .commit();
	
	    table.newAppend()
	        .appendFile(FILE_D)
	        .commit();
	
	    Iterable<DataFile> files = FindFiles.in(table).asOfTime(timestamp).collect();
	
	    Assert.assertEquals(pathSet(FILE_A, FILE_B), pathSet(files));
	  }
```

Reply:
Good catch. I should have looked at the logic more closely. I updated the test and fixed this. Thanks!